crates/spark/src/services/deposit.rs

```diff
@@ -198,7 +205,59 @@ impl<S: Signer> DepositService<S> {
                 vout,
             )
             .await?;
-        Ok(nodes)
+        self.collect_leaves(nodes).await
+    }
+
+    pub async fn collect_leaves(
+        &self,
+        nodes: Vec<TreeNode>,
+    ) -> Result<Vec<TreeNode>, ServiceError> {
+        let mut resulting_nodes = Vec::new();
+        for node in nodes.into_iter() {
+            if node.status != TreeNodeStatus::Available {
+                warn!("Leaf is not available: {:?}", node.clone());
+                // TODO: Handle other statuses appropriately.
+                resulting_nodes.push(node.clone());
+                continue;
+            }
+
+            let nodes = self.timelock_manager.extend_time_lock(&node).await?;
+
+            for n in nodes {
+                let node_id = n.id.clone();
+                if n.status != TreeNodeStatus::Available {
+                    warn!("Leaf resulting from extend_time_lock is not available: {n:?}",);
+                    // TODO: Handle other statuses appropriately.
+                    resulting_nodes.push(n);
+                    continue;
+                }
+
+                let transfer_res = self
+                    .transfer_service
+                    .transfer_leaves_to_self(
+                        vec![n],
+                        Some(PrivateKeySource::Derived(node.id.clone())),
+                    )
+                    .await;
+
+                let transfer = match transfer_res {
+                    Ok(transfer) => transfer,
+                    Err(e) => {
+                        if let ServiceError::TransferAlreadyClaimed = e {
+                            warn!("Transfer for leaf {} is already claimed", node_id);
+                            continue;
+                        }
+                        return Err(ServiceError::Generic(format!(
+                            "Failed to transfer leaves to self: {e:?}"
+                        )))?;
+                    }
+                };
+
+                resulting_nodes.extend(transfer.into_iter());
+            }
+        }
+
+        Ok(resulting_nodes)
```

Comment:
Looking at this function again, I think we can replace this with a call to `check_timelock_nodes` for the `Available` leaves. What do you think?

Reply:
I'll have to dive into it more. I created an issue for it: #213